<a href="https://colab.research.google.com/github/ivan74rus/DS/blob/main/ML_analis_klientov_banka.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Чтение-и-обработка-данных" data-toc-modified-id="Чтение-и-обработка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Чтение и обработка данных</a></span></li><li><span><a href="#Подготовка-даных-к-ML" data-toc-modified-id="Подготовка-даных-к-ML-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка даных к ML</a></span></li><li><span><a href="#разбивка-выборок-на-трен-обуч-и-тестовую" data-toc-modified-id="разбивка-выборок-на-трен-обуч-и-тестовую-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>разбивка выборок на трен обуч и тестовую</a></span></li><li><span><a href="#Применяем-методику-OHE-как-класс-из-sklearn" data-toc-modified-id="Применяем-методику-OHE-как-класс-из-sklearn-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Применяем методику OHE как класс из sklearn</a></span></li><li><span><a href="#Масштабирование" data-toc-modified-id="Масштабирование-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Масштабирование</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Обучение-моделей-без-учета-баланса-классов" data-toc-modified-id="Обучение-моделей-без-учета-баланса-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обучение моделей без учета баланса классов</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Используем-гипер-параметр--class_weight='balanced'" data-toc-modified-id="Используем-гипер-параметр--class_weight='balanced'-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Используем гипер параметр  class_weight='balanced'</a></span></li><li><span><a href="#Увеличение-выборки-upsampling" data-toc-modified-id="Увеличение-выборки-upsampling-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки upsampling</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Дополнительный-пункт." data-toc-modified-id="Дополнительный-пункт.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Дополнительный пункт.</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

In [ ]:
#устн нов версии , для колаба не нужно
#!pip install scikit-learn==1.1.3

In [ ]:
#подключить гугл диск
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# импорты
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import ( 
    f1_score,
    roc_auc_score,
    recall_score,
    precision_score
)
precision_score   
from sklearn.utils import shuffle
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder


## Подготовка данных

### Чтение и обработка данных

In [ ]:
#чтение файла и анализ
df =pd.read_csv('/content/drive/MyDrive/DS/Churn.csv' )
df.head(3)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Как видно,есть пропуске в столбце 'Tenure' - сколько лет человек является клиентом банка.     
Прежде чем обработать пропуски, посмотрим баланс класссов в целевом признаке по всему даатасету и по датасету с с пропусками.
Посмотрим корреляции.

In [ ]:
#баланс классов в целевом признаке в целом по дата сету
df['Exited'].mean()

0.2037

In [ ]:
# баланс классов в целевом признаке в строках с пропусками по колонке 'Tenure'
df[df['Tenure'].isna()]['Exited'].mean()


0.20132013201320131

In [ ]:
#посмотрим значения в колонке с пропусками
df['Tenure'].unique()

array([ 2.,  1.,  8.,  7.,  4.,  6.,  3., 10.,  5.,  9.,  0., nan])

In [ ]:
# кореляции
df.corr()

RowNumber  CustomerId  CreditScore       Age    Tenure  \
RowNumber         1.000000    0.004202     0.005840  0.000783 -0.007322   
CustomerId        0.004202    1.000000     0.005308  0.009497 -0.021418   
CreditScore       0.005840    0.005308     1.000000 -0.003965 -0.000062   
Age               0.000783    0.009497    -0.003965  1.000000 -0.013134   
Tenure           -0.007322   -0.021418    -0.000062 -0.013134  1.000000   
Balance          -0.009067   -0.012419     0.006268  0.028308 -0.007911   
NumOfProducts     0.007246    0.016972     0.012238 -0.030680  0.011979   
HasCrCard         0.000599   -0.014025    -0.005458 -0.011721  0.027232   
IsActiveMember    0.012044    0.001665     0.025651  0.085472 -0.032178   
EstimatedSalary  -0.005988    0.015271    -0.001384 -0.007201  0.010520   
Exited           -0.016571   -0.006248    -0.027094  0.285323 -0.016761   

                  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber       -0.009067       0.007246   0.000599        0.012044   
CustomerId      -0.012419       0.016972  -0.014025        0.001665   
CreditScore      0.006268       0.012238  -0.005458        0.025651   
Age              0.028308      -0.030680  -0.011721        0.085472   
Tenure          -0.007911       0.011979   0.027232       -0.032178   
Balance          1.000000      -0.304180  -0.014858       -0.010084   
NumOfProducts   -0.304180       1.000000   0.003183        0.009612   
HasCrCard       -0.014858       0.003183   1.000000       -0.011866   
IsActiveMember  -0.010084       0.009612  -0.011866        1.000000   
EstimatedSalary  0.012797       0.014204  -0.009933       -0.011421   
Exited           0.118533      -0.047820  -0.007138       -0.156128   

                 EstimatedSalary    Exited  
RowNumber              -0.005988 -0.016571  
CustomerId              0.015271 -0.006248  
CreditScore            -0.001384 -0.027094  
Age                    -0.007201  0.285323  
Tenure                  0.010520 -0.016761  
Balance                 0.012797  0.118533  
NumOfProducts           0.014204 -0.047820  
HasCrCard              -0.009933 -0.007138  
IsActiveMember         -0.011421 -0.156128  
EstimatedSalary         1.000000  0.012097  
Exited                  0.012097  1.000000

Баланс классов одинаковый, те наличие пропуска в колонке никак не влияет на уход/не уход клиента.    
Корреляция нулевая между Tenure и Exited
Пути решения:      
1) дропнуть эти строки и лишится 10% датасета для нашего исследования.     
2) заменить чем то  пропуски. Как вариант среднее/медиан.    
3) вобще выкинуть столбец этот.

Выбираем путь #2  с заменой пропусков медианой, т к лишаться на пустом месте 10% датасета для построения модели не хочется.

In [ ]:
# смотрим медианы и среднее
print ('медиана по столбцу "Tenure" ', df['Tenure'].median())
print ('среднее по столбцу "Tenure" ',df['Tenure'].mean())

медиана по столбцу "Tenure"  5.0
среднее по столбцу "Tenure"  4.997690023099769


Медиана и среднее равны, это говрит о том ,ч то нет каких то выбросов в данных.  Смело меняем пропуски на медиану.    
Затем еще приведем  к целочисленному типу.

In [ ]:
#замена NaN на медиану df['Tenure'].median()
df.loc[df['Tenure'].isna(),'Tenure'] = df['Tenure'].median()

In [ ]:
#изменения на целочисленный тип данн
df['Tenure'] = df['Tenure'].astype(int)
#проверка
df['Tenure'].head(3)

0    2
1    1
2    8
Name: Tenure, dtype: int64

### Подготовка даных к ML
1.Дропнем колонку с с индексом строки, именами клиентов и id клиента, они ничем не помогут  в ML     
2. Разобьем выборки на трен, валид и тестову 60/20/20%       
3.Применим технику кодирования One-Hot Encoding для преобразования категориальных признаков в численные.      
4. Мастштабируем численные признаки     
 


In [ ]:
# дроп колонки с индексом строки, именами клиентов и id клиента
df = df.drop(columns=['RowNumber','CustomerId','Surname'])

### разбивка выборок на трен обуч и тестовую


In [ ]:
# бьем на таргет и фьючи
target = df['Exited']
features = df.drop(columns='Exited')
target.shape
features.shape

(10000, 10)

In [ ]:
# разбиваем выборк на трен, (валид + тест)
features_train, features_valid_test, target_train, target_valid_test =\
             train_test_split(features, target, test_size=0.4, random_state=12345)
# разбиваем выборк на валид  и  тест
features_valid, features_test, target_valid, target_test =\
             train_test_split(features_valid_test, target_valid_test, test_size=0.5, random_state=12345)
# проверяем размерности
print ("features_train.shape", features_train.shape)
print ("features_valid.shape", features_valid.shape)
print ("features_test.shape", features_test.shape)

features_train.shape (6000, 10)
features_valid.shape (2000, 10)
features_test.shape (2000, 10)


In [ ]:
#проверим баланс классов. должен быть таким же как в исходном датасете
print ('target_train.mean', target_train.mean())
print ('target_valid.mean', target_valid.mean())
print ('target_test.mean', target_test.mean())

target_train.mean 0.19933333333333333
target_valid.mean 0.209
target_test.mean 0.2115


Все ок,  баланас классов соблюден.   
 Дисбаланс классов будет оказывать влияние на качество модели, хотя и не сильно критично.


### Применяем методику OHE как класс из sklearn

In [ ]:
#категориальные признаки для OHE кодирования

ohe_features = features_train.select_dtypes(include='object').columns.to_list()

print('Признаки для OHE кодировки',ohe_features)

Признаки для OHE кодировки ['Geography', 'Gender']


In [ ]:
# игнорируется ранее невстречающиеся значения признака (при transform)
encoder_ohe = OneHotEncoder(drop='first', handle_unknown='ignore', sparse=False)

# обучаем энкодер на заданных категориальных признаках тренировочной выборки
encoder_ohe.fit(features_train[ohe_features])

# добавляем закодированные признаки в features_train, valid, test
# encoder_ohe.get_feature_names_out() позволяет получить названия колонок
features_train[encoder_ohe.get_feature_names_out()] = encoder_ohe.transform(features_train[ohe_features])
features_valid[encoder_ohe.get_feature_names_out()] = encoder_ohe.transform(features_valid[ohe_features])
features_test[encoder_ohe.get_feature_names_out()] = encoder_ohe.transform(features_test[ohe_features])

# удаляем незакодированные категориальные признаки (изначальные колонки)
features_train = features_train.drop(ohe_features, axis=1)
features_valid = features_valid.drop(ohe_features, axis=1)
features_test = features_test.drop(ohe_features, axis=1)    
    
#проверям
print ('features_train.columns   ' , features_train.columns) 
print ('features_train.shape' ,features_train.shape)
print ('features_valid.shape' ,features_valid.shape)
print ('features_test.shape' ,features_test.shape)

features_train.columns    Index(['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Geography_Germany',
       'Geography_Spain', 'Gender_Male'],
      dtype='object')
features_train.shape (6000, 11)
features_valid.shape (2000, 11)
features_test.shape (2000, 11)


### Масштабирование

In [ ]:
# создаём скелер
scaler = StandardScaler()

#задаем численные признаки для масштабирования
#???HasCrCard   IsActiveMembe категориальный бинарный признак ,  не трансофрмируем  ???
#scaler_features = ['CreditScore', 'Age','Tenure','Balance','NumOfProducts', 'EstimatedSalary']

scaler_features = features_train.select_dtypes(exclude='object').columns.to_list()

# обучаем его на численных признаках тренировочной выборки
scaler.fit(features_train[scaler_features])

#масштабируем все выборки
features_train[scaler_features] = scaler.transform(features_train[scaler_features])
features_valid[scaler_features] = scaler.transform(features_valid[scaler_features])
features_test[scaler_features] = scaler.transform(features_test[scaler_features])

# проверяем
features_train.head(2)


CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
7479    -0.886751 -0.373192  1.082277  1.232271       -0.89156   0.642466   
3411     0.608663 -0.183385  1.082277  0.600563       -0.89156  -1.556504   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
7479       -1.055187        -0.187705          -0.572475         1.728977   
3411       -1.055187        -0.333945          -0.572475        -0.578377   

      Gender_Male  
7479     0.907278  
3411    -1.102198

**Вывод по п1 Подготовка данных**:
Данные были проанализированы.    
В столбце Tenure  пропуски, составляющие 10% от датасета были заменены на медиану, для того чтобы не терять данные для ML.  
Баланс классов целевого признака составляет 80/20 %, в дальнейшем придется с этим бороться.     
При подготовке данных к ML были дропнуты лишние  колонки  с именами, индексами и ID.      
Выборка разбита на обучающию и тестовую  в отношени60/20/20%.     
Применено кодирование OHЕ.    
И Мастштабирование  численных признаков.  


## Исследование задачи

### Обучение моделей без учета баланса классов

In [ ]:
#создадим дата сет для записил результатов 
d = {'no_balanced': [0,0,0], 'class_weight': [0,0,0] , 'upsampling': [0,0,0]}
result = pd.DataFrame(data=d, index=['LogisticRegression_F1', 'DecisionTreeClassifier_F1', 'RandomForestClassifier_F1'])
result

no_balanced  class_weight  upsampling
LogisticRegression_F1                0             0           0
DecisionTreeClassifier_F1            0             0           0
RandomForestClassifier_F1            0             0           0

In [ ]:
# LogisticRegression
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

#F1 метрика
print("F1:", f1_score(target_valid, predicted_valid))
#  AUC-ROC
probabilities_one_valid = model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )

#сохраняем метрику F1  в табличку результатов
result.loc['LogisticRegression_F1','no_balanced'] = f1_score(target_valid, predicted_valid) 

F1: 0.33108108108108103
AUC-ROC =  0.7586363333918061


In [ ]:
#DecisionTreeClassifier
f1=0
for i in range (1,20):
    model = DecisionTreeClassifier (max_depth=i)
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
            best_model = model
#метрика F1
print ('F1 = ',f1 , 'при max_depth=', max_depth)
print ()

#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )
print ()
print ()

#сохраняем метрику F1  в табличку результатов
result.loc['DecisionTreeClassifier_F1','no_balanced'] = f1

best_model_tree = best_model #для дальнейшего использования

F1 =  0.5658093797276854 при max_depth= 6

AUC-ROC =  0.8155286143758429




In [ ]:
%%time
# RandomForestClassifier
# сразу пройдемся циклом, поищем лучише гипер параметры


f1 = 0
for i in range(1,16,1):
    for j in range (30,101,10):
        model =  RandomForestClassifier(n_estimators=j,max_depth = i,random_state=12345)
        model.fit(features_train, target_train)
        predicted_valid = model.predict(features_valid)
        if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
            n_estimators = j
            best_model = model

#метрика F1            
print ('F1 = ',f1 , 'при max_depth=', max_depth, 'n_estimators=' , n_estimators)
print ()
print ()
#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )

#сохраняем метрику F1  в табличку результатов
result.loc['RandomForestClassifier_F1','no_balanced'] = f1

F1 =  0.5896656534954408 при max_depth= 12 n_estimators= 90


AUC-ROC =  0.8483976433440803
CPU times: user 46.6 s, sys: 145 ms, total: 46.8 s
Wall time: 47 s


In [ ]:
#результат промежуточный
result

no_balanced  class_weight  upsampling
LogisticRegression_F1         0.331081             0           0
DecisionTreeClassifier_F1     0.565809             0           0
RandomForestClassifier_F1     0.589666             0           0

**Вывод:**  ДеревоРешений и Случайный лес показывает гораздо лучший результат , чем логическая регрессия. 



## Борьба с дисбалансом

### Используем гипер параметр  class_weight='balanced'

In [ ]:
# LogisticRegression
model = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

#метрика F1 
print("F1:", f1_score(target_valid, predicted_valid))

#  AUC-ROC
probabilities_one_valid = model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )

#сохраняем метрику F1  в табличку результатов
result.loc['LogisticRegression_F1','class_weight'] = f1_score(target_valid, predicted_valid) 

F1: 0.4888507718696398
AUC-ROC =  0.7635692207187316


In [ ]:
#DecisionTreeClassifier
for i in range (1,20):
    model = DecisionTreeClassifier (max_depth=i,class_weight='balanced')
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
            best_model = model
#метрика F1
print ('F1 = ',f1 , 'при max_depth=', max_depth)
print ()

#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )
print ()
print ()

#сохраняем метрику F1  в табличку результатов
result.loc['DecisionTreeClassifier_F1','class_weight'] = f1

best_model_tree_class_weight = best_model #для дальнейшего использования

F1 =  0.5963791267305644 при max_depth= 5

AUC-ROC =  0.8310244134068074




In [ ]:
%%time
#37 sec
# RandomForestClassifier
# пройдемся циклом по важнейшим гиперпараметрам и найдем лучшию модель
f1 = 0
for i in range(1,16,1):
    for j in range (30,101,10):
        model =  RandomForestClassifier(n_estimators=j,max_depth = i, class_weight='balanced', random_state=12345)
        model.fit(features_train, target_train)
        predicted_valid = model.predict(features_valid)
        if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
            n_estimators = j
            best_model = model


# метрика F1
print ('F1 = ',f1 , 'при max_depth=', max_depth, 'n_estimators=' , n_estimators)
print ()

#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )
print ()
print ()

#сохраняем метрику F1  в табличку результатов
result.loc['RandomForestClassifier_F1','class_weight'] = f1

best_model_forest_class_weight = best_model #для дальнейшего испльзования

F1 =  0.6284403669724771 при max_depth= 9 n_estimators= 90

AUC-ROC =  0.8539475196438401


CPU times: user 48.4 s, sys: 171 ms, total: 48.6 s
Wall time: 51.7 s


In [ ]:
# результаты промежуточные
result

no_balanced  class_weight  upsampling
LogisticRegression_F1         0.331081      0.488851           0
DecisionTreeClassifier_F1     0.565809      0.596379           0
RandomForestClassifier_F1     0.589666      0.628440           0

**Вывод**
С гипер параметром  class_weight='balanced' Логическая регрессия показала уже удвол. результат. Случайный лес и дерево увеличили F1, добившись заданного порога больше 0.59 .

### Увеличение выборки upsampling
Преобразование проходит в несколько этапов:     
Разделить обучающую выборку на отрицательные и положительные объекты;     
Скопировать несколько раз положительные объекты;    
С учётом полученных данных создать новую обучающую выборку;    
Перемешать данные: идущие друг за другом одинаковые вопросы не помогут обучению.     

In [ ]:
# напишем функцию
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled


In [ ]:
# применяем функцию, т к соотновшение  в таргете у нас примерно 80 к 20, 
# то нужно повторить умножение положительных ответов 4х раза, чтобы уровнять  выборки

#получаем  новые датасеты
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

#проверяем размерность
print ('features_upsampled.shape',features_upsampled.shape)
print('target_upsampled.shape',target_upsampled.shape)

#проверяем баланс классов
print ('Баланс классов', target_upsampled.mean())

features_upsampled.shape (9588, 11)
target_upsampled.shape (9588,)
Баланс классов 0.49895702962035876


In [ ]:
# LogisticRegression обучаем на upsampled данных
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_upsampled, target_upsampled)
predicted_valid = model.predict(features_valid)

# метрика F1
print("F1:", f1_score(target_valid, predicted_valid))

#  AUC-ROC
probabilities_one_valid = model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )

#сохраняем метрику F1  в табличку результатов
result.loc['LogisticRegression_F1','upsampling'] = f1_score(target_valid, predicted_valid) 

F1: 0.4888507718696398
AUC-ROC =  0.763567708490857


In [ ]:
#DecisionTreeClassifier
for i in range (1,20):
    model = DecisionTreeClassifier (max_depth=i)
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
           
            best_model = model
#метрика F1
print ('F1 = ',f1 , 'при max_depth=', max_depth)
print ()

#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )
print ()

#сохраняем метрику F1  в табличку результатов
result.loc['DecisionTreeClassifier_F1','upsampling'] = f1

best_model_tree_upsampling = best_model

F1 =  0.6284403669724771 при max_depth= 9

AUC-ROC =  0.8539475196438401



In [ ]:
%%time
#49sec
# RandomForestClassifier обучим на upsampled данных
# пройдемся циклом по важнейшим гиперпараметрам и найдем лучшию модель
f1 = 0
for i in range(1,16,1):
    for j in range (30,101,10):
        model =  RandomForestClassifier(n_estimators=j,max_depth = i, random_state=12345)
        model.fit(features_upsampled, target_upsampled)
        predicted_valid = model.predict(features_valid)
        if f1 < f1_score (target_valid, predicted_valid):
            f1 = f1_score (target_valid, predicted_valid)
            max_depth = i
            n_estimators = j
            best_model = model


            
# метрика F1
print ('F1 = ',f1 , 'при max_depth=', max_depth, 'n_estimators=' , n_estimators)
print ()

#  AUC-ROC
probabilities_one_valid = best_model.predict_proba(features_valid)[:, 1]   #вероятности 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )
print ()
print ()

#сохраняем метрику F1  в табличку результатов
result.loc['RandomForestClassifier_F1','upsampling'] = f1


best_model_forest_upsampling = best_model #для дальнейшего использования


F1 =  0.6291179596174282 при max_depth= 9 n_estimators= 70

AUC-ROC =  0.8526303691650686


CPU times: user 1min 4s, sys: 214 ms, total: 1min 4s
Wall time: 1min 5s


In [ ]:
# итоговый результат сведеный в таблицу
result

no_balanced  class_weight  upsampling
LogisticRegression_F1         0.331081      0.488851    0.488851
DecisionTreeClassifier_F1     0.565809      0.596379    0.628440
RandomForestClassifier_F1     0.589666      0.628440    0.629118

**Вывод** модель случайного леса и дерева решений  гораздо лучше, против Логической регресии . Применение методов class_weight='balanced' и upsampling примерно равный результат дают.       
Лучшие модели DecisionTreeClassifier(max_depth= 9), на upsampling выборке и     
RandomForestClassifier(max_depth= 9 n_estimators= 70 , class_weigh='balanced')

## Тестирование модели


In [ ]:
#применяем лучшию модель на тестовой выборке
#best_model_forest_class_weight случ лес с гиперпараметром class_weight 
predicted_test = best_model_forest_class_weight.predict(features_test)

#метрика f1
print ('F1 = ',f1_score (target_test, predicted_test))

#метрика recall (полнота)
print('recall_score = ', recall_score(target_test, predicted_test) )

#метрика точность precision, сколько нашла отриц ответов ,пока искала положит
print ('precision_score = ', precision_score(target_test, predicted_test) )
#AUC-ROC площадь под кривой
probabilities_test = best_model_forest_class_weight.predict_proba(features_test)[:, 1]   

print ('AUC-ROC = ',roc_auc_score(target_test, probabilities_test) )

F1 =  0.6101694915254237
recall_score =  0.6382978723404256
precision_score =  0.5844155844155844
AUC-ROC =  0.8546751395278762


In [ ]:
# проверка на адекватность, все предсказания 1 или 0
predicted_const_1 = pd.Series(1, index=target_test.index)
predicted_const_0 = pd.Series(0, index=target_test.index)

#метрика f1
print ('Все 1,  F1 = ',f1_score (target_test, predicted_const_1))
print ('Все 0,  F1 = ',f1_score (target_test, predicted_const_0))


Все 1,  F1 =  0.3491539413949649
Все 0,  F1 =  0.0


Как видно из теста, наша модель показывать хорошие резузльтаты, по сравнению с константантными

**Вывод** Результат достигнут. F1>0.59   F1 =  0.61504.    
AUC-ROC  =  0.86096  говрит о том , что результат гораздо выше случайного (0.5)    
recall_score =   0.6572   показывает долю положительных ответов , которые спрогнозировала модель, впринципе, неплохой результат.   



**Общий вывод.**
1. Данные были узучены, пропуски в колонке с Tenure — сколько лет человек является клиентом банка заменены на медианы, чтобы не терять 10%  данных.
2. Баланс классов составляет 80/20. Корреляции слабые в датасете.   
3. Выбока разбита на обучающию, валидац, тестовую в отшении 0.6/ 0.2 / 0.2    
4. Обучены модели на обуч. выборке и применены на валидац ЛогРегресия, ДеревоРешений, СлучЛес. У дерева и Леса перебирались в цикле  важные гиперпараметры, для поиска наилучшей модели.Лучше всего себя показал случайны лес, ЛогРегрессия сильно хуже.
5. Для учета баланса классов был применен гиперпараметр class_weight='balanced'  и метод увелеичения редкого(положительного) класса  "upsampling". Так же для СлучЛеса и ДереваРешений циклом пербирались и находились  лучшие гиперпараметры. Два эти метода показали  примерно схожие увелечения метрики моделей. СлучайныйЛес и ДеревоРешений показали оптяь значительно лучший результат, по сравнению с ЛогРегрессией.
6. Была выбрана модель СлучайногоЛеса  с гипер параметр class_weight='balanced' и применена на тестовой выборке.   
Достигнут необходимый результат F1>0.59     
F1 =  0.61504.    
AUC-ROC  =  0.86096  говрит о том , что результат гораздо выше случайного (0.5)    
recall_score =   0.6572   показывает долю положительных ответов , которые спрогнозировала модель, впринципе хорошо.



## Дополнительный пункт.
улучшение модели  с гиперпараметрами class_weight='balanced' с помощью обьединения train и valid выборок

In [ ]:
# нашли лучшию модель ,в данном случае возьмем СлучайныйЛес с class_weight='balanced'  а теперь объеденим Трайн и Валид и обучим модель на них

#объединение трайн и валид
features_train_valid = pd.concat([features_train, features_valid], ignore_index=True)
target_train_valid = pd.concat([target_train, target_valid], ignore_index=True)

# обучаем лучшию модель на объеденнных данных, выберем случайный лес
model =  best_model_forest_class_weight
model.fit(features_train_valid, target_train_valid)
predicted_valid = model.predict(features_valid)

# метрики 
print ('F1 на= ',f1_score (target_valid, predicted_valid))

probabilities_valid = model.predict_proba(features_valid)   #вероятности 1
probabilities_one_valid = probabilities_valid[:, 1] 
print ('AUC-ROC = ',roc_auc_score(target_valid, probabilities_one_valid) )

best_model_train_valid = model  # для дальнейшего использования


F1 на=  0.7464324917672888
AUC-ROC =  0.9372138108747331


In [ ]:
#применяем лучшию модель на тестовой выборке

predicted_test = best_model_train_valid.predict(features_test)

#метрика f1
print ('F1 = ',f1_score (target_test, predicted_test))

#AUC-ROC площадь под кривой
probabilities_test = best_model_train_valid.predict_proba(features_test)[:, 1]   

print ('AUC-ROC = ',roc_auc_score(target_test, probabilities_test) )

F1 =  0.6150442477876107
AUC-ROC =  0.8609623263490693


На тестовой выборке нельзя продолжать испытывать модели, но я сделал. И получил одинаковые результаты на Случайный лес с гиперпараметром  class_weight='balanced'  и лучшими гиперпараметрами , найденными в цикле , обученной на   объединененной выборке train и valid.

И  на Дереве решений   обученным на upsampling выборкой с циклом для поиска лучших гипер параметров .




прошлый лучший результат на  тестов выборке.
F1 =  0.6037735849056605          
F1 = 0.6203090507726269 без масштабир    
F1 =  0.6150442477876107 масшт     